### ISSUE TITLE
Proposal: Manage error spam in the hostmetrics process scraper

### ISSUE THREAD

**Issue by braydonk** [open-telemetry/opentelemetry-collector-contrib#34988]
### Component(s)

receiver/hostmetrics

### Is your feature request related to a problem? Please describe.

The process scraper in hostmetrics is in a particularly strange place when it comes to error handling. The current default behaviour is to scrape all pieces of information from every process, and when a piece of information from a particular process can't be read, produce an error. These errors are collected and returned as a multi-error from the scraper, which the collector's scrapercontroller subsequently logs at ERROR level.

This would be fine if it weren't for the fact that numerous checks on specific processes are essentially guaranteed to fail. There are some processes where there isn't a name to read, or certain pieces of data are by design inaccessible and will always produce an error. This means that, in almost all deployments, the process scraper will be logging some kind of error every scrape, the majority of which are not actionable and are simply noise.

This has been a problem for the community for some time. Over the past 2 years, various `mute_*_error` config options have been added to opt-out of specific errors being included in the error messages (and [I was the one who started that](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/7176), sorry :smiling_face_with_tear: ).

### Describe the solution you'd like

I think the process scraper should take control of this by no longer returning errors reading specific parts of the process back to the scrapercontroller. Errors that actually cause a failed scrape are the only errors that should be returned to scrapercontroller.

The logging of errors reading data from specific processes should actually be handled by the processscraper. By default, it would log the combined process errors at DEBUG level. A configuration option would allow a user to log these errors as ERROR instead of DEBUG. This way, seeing this data is opt-in, either by enabling debug logging in the collector or flipping this configuration switch.

The existing `mute_*_error` configuration options would be deprecated and would have no longer have any effect once this change was implemented. They would be kept around for a few releases with deprecation warnings to not break user configuration without notice.

### Describe alternatives you've considered

As stated above, what currently what exists are numerous opt-out flags for specific types of errors. This is not ideal; it doesn't cover all error types, and has slowly encouraged a balooning of options that users have to know.

[This PR](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/34981) adds an option that would mute any errors being logged. This is a step in the right direction, but I think it would be better to go all the way into making this switch entirely in how the scraper handles these types of errors.

I tried to [pitch an idea](https://github.com/open-telemetry/opentelemetry-collector/issues/8293) to the upstream scraper libraries that would allow scrapers to return errors that are wrapped in a level directive, meaning when the `scrape` function returns an error it could be in control of what level the scrapercontroller logs those errors at. However, I now think this probably wouldn't be that useful as a library feature; I think this is a very special problem that only the process scraper is currently dealing with.

### Additional context

This is currently the most prominent issue brought up in the hostmetrics component's list of issues. Here are just a few examples of users who have been experiencing this issue:
- #20435
- #34557
- #30359
- #21665
There are even more if you keep searching. It's been a constant source of pain for users, and a timesink for maintainers.

**Comment by rogercoll** [comment]
> I think the process scraper should take control of this by no longer returning errors reading specific parts of the process back to the scrapercontroller. 

+1 The key point is that we're incorrectly flagging statements as errors when they aren't. Even if you run the collector with root privileges, the processor will log this error for all kernel processes: 

```
error   scraperhelper/scrapercontroller.go:197  Error scraping metrics      {"kind": "receiver", "name": "hostmetrics/producer", "data_type": "metrics", "error": "error reading process executable for pid 2: readlink /proc/2/exe: no such file or directory; error reading process executable for pid 3:
...
```

The error explains that the scraper could not read from a process's `/proc/<PID>/exe` file, but kernel threads do not correspond to an executable file on disk, there is no /proc/[PID]/exe for these processes. Having no "EXE" file is expected, no error should be logged.

```
(root) $  cat /proc/2/status | rg Name
Name:   kthreadd
(root) $ cat /proc/2/exe
cat: /proc/2/exe: No such file or directory
```

For the configuration option, it might be useful to leverage existing configurations from other components. What do you think about reusing the `error_mode` setting, as it's implemented in the [transform processor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/transformprocessor#config)?

**Comment by braydonk** [comment]
> The error explains that the scraper could not read from a process's /proc/<PID>/exe file, but kernel threads do not correspond to an executable file on disk, there is no /proc/[PID]/exe for these processes. Having no "EXE" file is expected, no error should be logged.

Yes this is the most frustrating one. This is along with various special processes on Windows are all but guaranteed to produce some manner of errors every scrape, so this problem affects every platform.

I'd thought about trying to address all of these edge cases, but there are always scenarios where people introduce their own programs where these expected errors are produced as well (I can't remember the specifics, but there was one issue where a certain Oracle DB program on Windows didn't have a "command" and produced an expected "error reading command line" error).

> What do you think about reusing the error_mode setting, as it's implemented in the [transform processor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/transformprocessor#config)?

Yeah that seems like a great fit. I'm not sure if it will look odd to take that directly from the `pkg/ottl` in this receiver that is unrelated to OTTL, but that is what I will do in my POC.

**Comment by crobert-1** [comment]
_Looks like there's a general agreement on a path forward, and this was filed by a code owner. Removing `needs triage` with the understanding that some details may still need to be ironed out._

**Comment by dmitryax** [comment]
I like the approach. I've been thinking about a pretty similar solution. We need to think how to make the new `scraperhelper`'s `level` option applied on the receivers in a consistent way with less code duplication. Should be similar to what we do in the exporterhelper. Even if some receivers don't expose that config option, it's still fine to have DEBUG level by default.
